plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/protobufnative/PulsarProtobufNativeRowDecoderFactory.java

```diff
@@ -0,0 +1,169 @@
+                                parseProtobufTrinoType(field), field.getType().toString(), null, false, false,
+                                handleKeyValueType, new PulsarColumnMetadata.DecoderExtraInfo(field.getName(),
+                                null, null))
+
+                ).collect(toList());
+
+        return columnMetadata;
+    }
+
+    private Type parseProtobufTrinoType(Descriptors.FieldDescriptor field)
+    {
+        //parse by proto JavaType
+        Descriptors.FieldDescriptor.JavaType type = field.getJavaType();
+        Type dataType;
+        switch (type) {
+            case BOOLEAN:
+                dataType = BooleanType.BOOLEAN;
+                break;
+            case BYTE_STRING:
+                dataType = VarbinaryType.VARBINARY;
+                break;
+            case DOUBLE:
+                dataType = DoubleType.DOUBLE;
+                break;
+            case ENUM:
+            case STRING:
+                dataType = createUnboundedVarcharType();
+                break;
+            case FLOAT:
+                dataType = RealType.REAL;
+                break;
+            case INT:
+                dataType = IntegerType.INTEGER;
+                break;
+            case LONG:
+                dataType = BigintType.BIGINT;
+                break;
+            case MESSAGE:
+                Descriptors.Descriptor msg = field.getMessageType();
+                if (field.isMapField()) {
+                    //map
+                    TypeSignature keyType =
+                            parseProtobufTrinoType(msg.findFieldByName(PulsarProtobufNativeColumnDecoder.PROTOBUF_MAP_KEY_NAME)).getTypeSignature();
+                    TypeSignature valueType =
+                            parseProtobufTrinoType(msg.findFieldByName(PulsarProtobufNativeColumnDecoder.PROTOBUF_MAP_VALUE_NAME)).getTypeSignature();
+                    return typeManager.getParameterizedType(StandardTypes.MAP,
+                            ImmutableList.of(TypeSignatureParameter.typeParameter(keyType),
+                                    TypeSignatureParameter.typeParameter(valueType)));
+                }
+                else {
+                    //row
+                    dataType = RowType.from(msg.getFields().stream()
+                            .map(rowField -> new RowType.Field(Optional.of(rowField.getName()),
+                                    parseProtobufTrinoType(rowField)))
+                            .collect(toImmutableList()));
+                }
+                break;
+            default:
+                throw new RuntimeException("Unknown type: " + type.toString() + " for FieldDescriptor: " + field.getName());
+        }
+        //list
+        if (field.isRepeated() && !field.isMapField()) {
+            dataType = new ArrayType(dataType);
+        }
+
+        return dataType;
+    }
+
+    private static final Logger log = Logger.get(PulsarProtobufNativeRowDecoderFactory.class);
```

Comment:
we define these loggers at the top of the class (convention)

Reply:
updated